design_docs/generalized_horizontal_interpolation.md

```diff
@@ -0,0 +1,95 @@
+
+<h2>Design solution: Cartesian input mesh
+Date last modified: 2017/01/19
+Contributors: Xylar Asay-Davis</h2>
+
+Currently, lonCell and latCell are passed as input arguments.  For efficiency, it would make more sense to use xCell, yCell, zCell rather than lonCell and latCell.  (Currently, fields equivalent to xCell, yCell and zCell are being computed from lonCell and latCell in the nearest neighbor code.)  Moreover, to support more generality in the interpolation algorithm (see next design solution), it may be best to pass an xarray data set containing xCell, yCell and zCell but also other mesh variables needed for linear interpolation.
+
+<h2>Design solution: First-order interpolation
+Date last modified: 2017/01/19
+Contributors: Xylar Asay-Davis</h2>
+
+Linear interpolation will require locations of MPAS mesh cells, vertices and a list of cellsOnVertex.  Therefore, it seems simplest if the routine for initializing interpolation takes an MPAS mesh data set as an input argument, from which it can read whichever geometric information is needed for interpolation.
+The output of the initialization routine (for either linear or nearest-neighbor interpolation) will be a 2D array of MPAS cell indices and corresponding interpolation weights.  The first dimension in each array corresponds to the output interpolation points.  The second index is over input cells neighboring the output point (either a single entry for nearest-neighbor or vertexDegree entries for linear).
+The algorithm for linear interpolation will be as follows:
+Store the shape of the array of output points
+Use a KD tree method (the same currently used to find nearest neighboring cells) to find nearest neighboring vertices to each output point
+For each output point at (xp, yp, zp), the interpolation weight for each neighboring cell is given by barycentric weights (e.g. http://answers.unity3d.com/questions/383804/calculate-uv-coordinates-of-3d-point-on-plane-of-m.html).  If only 2 neighboring cells are available, linear interpolation based on distance between the cells should be used instead.  If only one neighboring cell exists, the weight of that cell is (obviously) 1.  Note: if we want to support vertexDegree = 4, we would need to either use Wachspress coordinates or divide each quadrilateral into 2 triangles.
+Store the indices and weights for interpolating arbitrarily many fields
+For each field, interpolation is as simple as indexing the field at the given indices, multiplying by the weights, summing over neighbors, and reshaping the field to conform to the shape stored in step 1.
+It will likely be desirable to create an Interpolator class that keeps track of the indices, weights and output shape so these don't need to be stored or passed in later as input arguments.  This approach is also consistent with how scipy handles interpolation (https://docs.scipy.org/doc/scipy-0.18.1/reference/interpolate.html)
+
+<h2>Design solution: Interpolation should handle periodic boundaries
```

Comment:
I'm worried that if we exclude periodic boundaries now, we'll never get to them.  There's a test case I use frequently (`sub_ice_shelf_2D`) that has periodic boundaries and the fact that our paraview viz tools don't support it has actually kept me from making headway on that test case several times.

At the same time, I will only handle periodic boundaries if it's not a major time sink for this project.  So maybe I'll make it a consideration rather than a requirement.